### DISCUSSION TITLE
Getting categories and models with overridden display

### DISCUSSION THREAD

**Discussion by AntanasBuk** [iTwin/itwinjs-core#5772] [Q&A]
Hey,

I am trying to make a button that would clear all overrides (hidden/isolated) made to elements, categories and models in the view. If there is nothing to clear, the button would be hidden.
Now doing that for individual elements is easy, since all overridden elements can be found under neverDrawn or alwaysDrawn properties. 
Unfortunately though I can not say the same for overridden categories and models since all I could find in frontend-core api were sets of category and model selectors containing currently displayed items. 

Is there a way of retrieving a set of in some way overridden categories and models without using queries? Cause I imagine using queries every time something changes in the view would be very costly.

**Comment by pmconne** [comment]
In the context of what app?

**Reply by AntanasBuk** [reply]
This would be a fix for AppUI's clearHideIsolateEmphasizeElements

**Reply by pmconne** [reply]
Do you have a link to the issue describing the thing that needs fixing?
Is AppUI also the one responsible for applying the overrides that it's supposed to clear? Does it do so exclusively using the `EmphasizeElements` API?

**Reply by AntanasBuk** [reply]
https://github.com/iTwin/appui/issues/264 the issue is only on elements, I only discovered issues with categories and models later witch I remarked in the comments bellow.
AppUI currently tracks overridden categories and objects (which is why it gets desynced with presentation's Models Tree when changes happen on their side), and makes use of `EmphasizeElements` API for elements. Yes, there is functionality in AppUI to override selected elements/categories/models (currently I only fixed the original issue and issues related to desyncing of element display state).

**Reply by pmconne** [reply]
The visibility and appearance of elements, models, and categories ("features") can be overridden in a [variety of ways](https://www.itwinjs.org/learning/display/symbologyoverrides/) for a  variety of purposes. The viewport displays the union of all of those overrides. A tool that arbitrarily queries and resets all overrides without consideration for the purpose for which each override was applied is not very useful.

It sounds like the issue is that the tool provided by appui needs to synchronize with overrides applied by the presentation components. Have you talked to the maintainers of those components, or looked at their source code?

**Reply by AntanasBuk** [reply]
The tool is for hiding, isolating and emphasizing elements though only hide/show is relevant in presentation's case. I've talked to presentation team and they said that they are solving a different problem - they don't need to know if some feature is hidden if the tree didn't load it yet - though I'll have to ask how they handle visibility changes from outside. I skimmed through their code but couldn't make out how exactly they detect what and where changed.
It just seemed strange that there are at least three different APIs for relatively similar use cases of overriding and tracking overridden features.

**Comment by NancyMcCallB** [comment]
@pmconne There's no good reason for AppUi to own these tools at all, but they were developed as part of the base UI code years ago. I think we should deprecate the ones in AppUi, but I don't know where to move the implementations.
